### DISCUSSION TITLE
Pre-registering of buffers.

### DISCUSSION THREAD

**Discussion by andrii0lomakin** [bytedance/monoio#66] [Q&A] [answered]
Hi guys.
Do you have any means on the library level to pre-register buffers that are used to read/write data from/to io_uring?

**Comment by dyxushuai** [comment, marked as answer]
Yes, We have some ideas about the IO API. Which provides methods to manipulate the buffer directly. e.g.
```rust
trait BufRead {
	async fn next(&mut self, n: usize) ->  &[u8];
	async fn peek(&mut self, n: usize) -> &[u8];
	async fn advance(&mut self, n: usize);
}
```
We just expose the buffer operations and we can use pre-register buffers as one of our implements.

**Reply by andrii0lomakin** [reply]
Thank you. What about mutable form? Will it be possible not only to read but write in such a buffer?

**Reply by dyxushuai** [reply]
My mistake. This example is not the actual trait. We are still discussing this. Both `BufRead` and `BufWrite` all are required.
